overlord/state/taskrunner.go

```diff
@@ -243,26 +243,17 @@ func (r *TaskRunner) run(t *Task) {
 				t.At(timeNow().Add(x.After))
 			}
 		case nil:
-			var next []*Task
 			switch t.Status() {
 			case DoingStatus:
 				t.SetStatus(DoneStatus)
-				fallthrough
-			case DoneStatus:
-				next = t.HaltTasks()
 			case AbortStatus:
 				// It was actually Done if it got here.
 				t.SetStatus(UndoStatus)
 				r.state.EnsureBefore(0)
 			case UndoingStatus:
 				t.SetStatus(UndoneStatus)
-				fallthrough
-			case UndoneStatus:
-				next = t.WaitTasks()
-			}
-			if len(next) > 0 {
-				r.state.EnsureBefore(0)
 			}
+			r.state.EnsureBefore(0)
```

Comment:
These changes look fine but I think it would be good to have a comment explaining that we run EnsureBefore unconditionally here to also consider cleanup handlers.

Reply:
Oh, yes, I'll add comments, and also the unit tests need to be updated, but I wanted to get Samuele's feedback first, in order to avoid wasting time in case this turns out to be the wrong solution.

Reply:
this change is fine but creates more unproductive Ensure cycles in some cases,  a different approach would be to return a flag from Task.SetStatus that tells us if the Change status itself has changed or something like that. So the condition would become:

len(next) > 0 || taskChangeStatusChanged